sklearn/cluster/_k_means_fast.pyx

```diff
@@ -7,155 +7,256 @@
 #         Lars Buitinck
 #
 # License: BSD 3 clause
-#
-# cython: boundscheck=False, wraparound=False, cdivision=True
 
-from libc.math cimport sqrt
 import numpy as np
-import scipy.sparse as sp
 cimport numpy as np
 cimport cython
 from cython cimport floating
+from libc.math cimport sqrt
+
+from ..utils.extmath import row_norms
+
+
+np.import_array()
 
-from ..utils.sparsefuncs_fast import assign_rows_csr
-from ..utils._cython_blas cimport _dot
 
 ctypedef np.float64_t DOUBLE
 ctypedef np.int32_t INT
 
 
-np.import_array()
+cdef floating _euclidean_dense_dense(floating* a,
+                                     floating* b,
+                                     int n_features,
+                                     bint squared) nogil:
+    """Euclidean distance between a dense and b dense"""
+    cdef:
+        int i
+        int n = n_features // 4
+        int rem = n_features % 4
+        floating result = 0
+
+    # We manually unroll the loop for better cache optimization.
+    for i in range(n):
+        result += ((a[0] - b[0]) * (a[0] - b[0])
+                  +(a[1] - b[1]) * (a[1] - b[1])
+                  +(a[2] - b[2]) * (a[2] - b[2])
+                  +(a[3] - b[3]) * (a[3] - b[3]))
+        a += 4; b += 4
+
+    for i in range(rem):
+        result += (a[i] - b[i]) * (a[i] - b[i])
+
+    return result if squared else sqrt(result)
+
+
+cdef floating _euclidean_sparse_dense(floating[::1] a_data,
+                                      int[::1] a_indices,
+                                      floating[::1] b,
+                                      floating b_squared_norm,
+                                      bint squared) nogil:
+    """Euclidean distance between a sparse and b dense"""
+    cdef:
+        int nnz = a_indices.shape[0]
+        int i
+        floating tmp = 0.0
+        floating result = 0.0
+
+    for i in range(nnz):
+        tmp = a_data[i] - b[a_indices[i]]
+        result += tmp * tmp - b[a_indices[i]] * b[a_indices[i]]
```

Comment:
you meant `a_data[i]**2 - 2 * a_data[i] * b[a_indices[i]]`. It's mathematically the same nut not floating point arithmetically the same. Basically you have a much bigger loss of precision when squaring before taking the difference than the opposite. It's because when squaring, you'd need twice as more digits to store the result, but in fixed precision, half of the digits are truncated.

However you're right that the triple indirect memory acces `b[a_indices[i]]` is meh. I'm not sure that gcc does put it in a temp on its own.